libraries/standard/http/src/http_client.c

```diff
@@ -245,6 +245,986 @@ static int findHeaderValueParserCallback( http_parser * pHttpParser,
+ * - #HTTP_PARSER_INTERNAL_ERROR
+ */
+static HTTPStatus_t processHttpParserError( http_parser * pHttpParser );
+
+/*-----------------------------------------------------------*/
+
+static void processCompleteHeader( HTTPParsingContext_t * pParsingContext )
+{
+    HTTPResponse_t * pResponse = NULL;
+
+    assert( pParsingContext != NULL );
+    assert( pParsingContext->pResponse != NULL );
+
+    pResponse = pParsingContext->pResponse;
+
+    /* A header is complete when both the last header field and value have been
+     * filled in. */
+    if( ( pParsingContext->pLastHeaderField != NULL ) &&
+        ( pParsingContext->pLastHeaderValue != NULL ) )
+    {
+        /* Increase the header count. */
+        pResponse->headerCount++;
+
+        LogDebug( ( "Response parsing: Found complete header: "
+                    "HeaderField=%.*s, HeaderValue=%.*s",
+                    pParsingContext->lastHeaderFieldLen,
+                    pParsingContext->pLastHeaderField,
+                    pParsingContext->lastHeaderValueLen,
+                    pParsingContext->pLastHeaderValue ) );
+
+        /* If the application registered a callback, then it must be notified. */
+        if( pResponse->pHeaderParsingCallback != NULL )
+        {
+            pResponse->pHeaderParsingCallback->onHeaderCallback(
+                pResponse->pHeaderParsingCallback->pContext,
+                ( const uint8_t * ) pParsingContext->pLastHeaderField,
+                pParsingContext->lastHeaderFieldLen,
+                ( const uint8_t * ) pParsingContext->pLastHeaderValue,
+                pParsingContext->lastHeaderValueLen,
+                pResponse->statusCode );
+        }
+
+        /* Prepare the next header field and value for the first invocation of
+         * httpParserOnHeaderFieldCallback() and
+         * httpParserOnHeaderValueCallback(). */
+        pParsingContext->pLastHeaderField = NULL;
+        pParsingContext->lastHeaderFieldLen = 0;
+        pParsingContext->pLastHeaderValue = NULL;
+        pParsingContext->lastHeaderValueLen = 0;
+    }
+}
+
+/*-----------------------------------------------------------*/
+
+/* MISRA directive 4.6 violation is for using primitive type int, instead of
+ * a typedef which denotes the signedness and size of the type. This violation
+ * is suppressed because this callback follows the function signature for the
+ * http_cb type in http-parser. */
+
+/* MISRA directive 8.13 flags that pHttpParser should be a const parameter since
+ * it does not change. This rule is suppressed because http-parser defines
+ * their callback function signatures without the const. */
+/* coverity[misra_c_2012_directive_4_6_violation] */
+/* coverity[misra_c_2012_rule_8_13_violation] */
+static int httpParserOnMessageBeginCallback( http_parser * pHttpParser )
+{
+    HTTPParsingContext_t * pParsingContext = NULL;
+    HTTPResponse_t * pResponse = NULL;
+
+    assert( pHttpParser != NULL );
```

Comment:
don't we need to assert pHttpParser->data != NULL as well here?

Reply:
yes thanks